src/use-idle/use-idle.ts

```diff
@@ -0,0 +1,73 @@
+import { IdleController } from './idle-controller'
+
+const defaultEvents = ['mousemove', 'mousedown', 'resize', 'keydown', 'touchstart', 'wheel'];
+const oneMinute = 60e3;
+
+export const useIdle = (controller: IdleController, ms: number = oneMinute, initialState: boolean = false, events: string[] = defaultEvents) => {
+  let isIdle = initialState;
+  let timeout = setTimeout(() => {
+    isIdle = true
+    dispatchAway();
+  }, ms);
+
+  const method = (methodName: string): Function => {
+    const method = (controller as any)[methodName]
+    if (typeof method == 'function') {
+      return method
+    }
+    throw new Error(`undefined method "${methodName}"`)
+  }
+
+  const dispatchAway = () => {
+    controller.isIdle = true;
+    controller.away && method('away').call(controller);
+  }
+
+  const dispatchBack = () => {
+    controller.isIdle = false;
+    controller.back && method('back').call(controller);
+  }
+
+  const onEvent = () => {
+    if (isIdle) dispatchBack();
+
+    isIdle = false;
+    clearTimeout(timeout);
+
+    timeout = setTimeout(() => {
+      isIdle = true
+      dispatchAway();
+    }, ms);
+  }
+
+  const onVisibility = () => {
+    if (!document.hidden) onEvent();
+  };
+
+  if (isIdle) {
+    dispatchAway();
+  } else {
+    dispatchBack();
+  }
+
+  Object.assign(controller, {
+    observe() {
+      events.forEach(event => {
+        window.addEventListener(event, onEvent);
+      });
+      document.addEventListener('visibilitychange', onVisibility);
+    },
+    unObserve() {
+      events.forEach(event => {
+        window.removeEventListener(event, onEvent);
+      });
+      document.removeEventListener('visibilitychange', onVisibility);
+    },
+    disconnect() {
+      this.unObserve();
+      controller.disconnect();
```

Comment:
I had to do something like this to avoid having an infinite loop on the disconnect function

```js
  const controllerDisconnect = controller.disconnect

  Object.assign(controller, {
    disconnect() {
      this.unObserve()
      controllerDisconnect()
    },
  })
```

Reply:
Ahh, that makes sense. You are "saving" the old `disconnect` function before reassigning the new one with `Object.assign` to later call the old one again. I will update it 😊